examples/dynamoathenamigration/migration.go

```diff
@@ -486,18 +489,252 @@ func (t *task) downloadFromS3(ctx context.Context, key string) (*os.File, error)
 	}
 	path := path.Join(dir, originalName)
 
-	f, err := os.Create(path)
+	originalFile, err := os.Create(path)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	if _, err := t.s3Downloader.Download(ctx, f, &s3.GetObjectInput{
+	if _, err := t.s3Downloader.Download(ctx, originalFile, &s3.GetObjectInput{
 		Bucket: aws.String(t.Bucket),
 		Key:    aws.String(key),
 	}); err != nil {
-		f.Close()
+		originalFile.Close()
+		return nil, trace.Wrap(err)
+	}
+
+	defer originalFile.Close()
+
+	sortedFile, err := createSortedExport(originalFile, dir, 1024*1024*t.MaxMemoryUsedForSortingExport)
+	if err != nil {
+		sortedFile.Close()
+		return nil, trace.Wrap(err)
+	}
+	return sortedFile, nil
+}
+
+type eventWithTime struct {
+	rawLine   json.RawMessage
+	eventDate string
+}
+
+// createSortedExport reads a large json.gz export file, which cannot fit entirely
+// into memory and splits it into multiple smaller sorted (in memory) files.
+// Afterwards, it merges these smaller files into a final sorted output.
+// The function returns json file which is sorted by createdAtDate ascending.
+func createSortedExport(inFile *os.File, dir string, maxSize int) (*os.File, error) {
+	tmpSortedFilesDir := path.Join(dir, fmt.Sprintf("sort-%s", uuid.NewString()))
+	if err := os.Mkdir(tmpSortedFilesDir, 0o755); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	gzipReader, err := gzip.NewReader(inFile)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer gzipReader.Close()
+
+	dec := json.NewDecoder(gzipReader)
+	events := make([]eventWithTime, 0, 10000)
```

Comment:
should we start with 0? I have selected it arbitrary, because it will be hard to tell how much will fit into single chunk. We could estimate it somehow by using maxMemory passed and diving it by average event size. Do you think it will be better?

Reply:
https://github.com/gravitational/teleport/pull/29627/commits/f47e35ce81378849d3a02704f0ff1c493793a636
in i have tried to do some estimates, tomorrow I will try it again on large export file. let me know if this is better.